### PR TITLE
fix: preserve redirect through Singpass login

### LINKED ISSUE #2375: (warden) landingUrl in singpassLoginSchema is validated but never persisted or used for post-login redirect
# `landingUrl` in `singpassLoginSchema` is validated but never persisted or used for post-login redirect

**Skill:** code-review

**Severity:** 🟡 MEDIUM

**ID:** `PRL-L2U`

**Confidence:** high

**Location:** `apps/studio/src/schemas/auth/singpass.ts:5-6`

## Description

The new `singpassLoginSchema` accepts and validates a `landingUrl` (via `callbackUrlSchema`), but in `singpass.router.ts` the login mutation only logs the value and never stores it in the session state. The `SessionData` interface has no `landingUrl` field, so the callback procedure has no access to the original URL and unconditionally returns `redirectUrl: DASHBOARD`. Any caller passing a non-dashboard `landingUrl` to trigger a post-login redirect will always be sent to the dashboard instead of their intended destination.

## Verification

Traced the full flow: (1) `singpassLoginSchema` at line 5 defines `landingUrl: callbackUrlSchema` which transforms input into a `URL` object; (2) `singpass.router.ts` login mutation destructures `landingUrl` from input (line 20) and only logs it (line 37); (3) `set(ctx.session, 'singpass.sessionState', { ...session, userId, verificationToken })` at lines 44-48 of the router never includes `landingUrl`; (4) `session.ts` `SessionData.singpass.sessionState` has no `landingUrl` field; (5) the callback procedure at line 191 of the router hardcodes `redirectUrl: DASHBOARD`. No other code path stores or forwards `landingUrl`.

### PR DESCRIPTION
**Pull request by Benjam11n**
## Problem

Users signing in with Singpass are redirected to the dashboard instead of their original destination.

Closes #2375

## Solution

The original URL is now passed from the OTP page to the Singpass page, stored temporarily in the session, and returned after authentication.

If no URL is available, users are redirected to the dashboard.

**Breaking Changes**

- [ ] Yes - this PR contains breaking changes
- [x] No - this PR is backwards compatible

**Features**:

- N/A

**Improvements**:

- N/A

**Bug Fixes**:

- Preserve the original destination through the Singpass authentication flow.

## Before & After Screenshots

**BEFORE**:

N/A

**AFTER**:

N/A

## Tests

Updated `apps/studio/src/server/modules/auth/singpass/__tests__/singpass.router.test.ts` to verify that:

- The landing URL is stored in the Singpass session.
- The stored landing URL is returned after successful authentication.

**Manual Verification Steps**:

- [ ] Open `/sign-in?callbackUrl=%2Fsites%2F1` while logged out.
- [ ] Complete OTP and Singpass authentication.
- [ ] Confirm that you are redirected to `/sites/1`.

**New scripts**:

- N/A

**New dependencies**:

- N/A

**New dev dependencies**:

- N/A

<details><summary><h3>Greptile Summary</h3></summary>

This PR fixes users being redirected to the dashboard after Singpass login instead of their original destination. The original URL is now captured at the OTP step, forwarded as a query param to the Singpass page, validated server-side with the existing `callbackUrlSchema` (same-origin guard), stored in the Singpass session, and returned as `redirectUrl` after successful authentication.

- **`getRedirectUrl` bug fix**: Removes a `decodeURIComponent` call that was applied on top of Next.js's own query-param decoding, which was mangling percent-encoded characters (e.g. `%26` → `&`) inside the callback URL.
- **Session plumbing**: `landingUrl` is persisted in the Singpass `sessionState` during `login` and read back in `callback`; falls back to `DASHBOARD` when absent.
- **React correctness**: `EnforceLoginStatePageWrapper` and `DefaultTrpcError` both move their redirect side-effects from render time into `useEffect`, fixing React anti-patterns in the existing code.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

- Safe to merge. The redirect chain is validated at every boundary — client-side via `callbackUrlSchema` before the Singpass mutation, server-side in the schema for `singpassLoginSchema`, and again in `SingpassCallback` before the final navigation — so no open-redirect exposure is introduced.
- The change is well-scoped: a new session field, a forwarded query param, and the removal of a double-decode bug. All URL validation passes through the existing `callbackUrlSchema` with its same-origin whitelist. Unit and e2e tests cover the key path including the encoded-ampersand edge case.
- The error-handling path in `DefaultTrpcError.tsx` silently swallows a navigation failure and still resets the error boundary, which may leave the UI in an unclear state; worth a quick look but not blocking.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/studio/src/utils/url.ts | Removes redundant `decodeURIComponent` call from `getRedirectUrl`; Next.js `router.query` already decodes query params, so the old code caused double-decoding which mangled percent-encoded characters (e.g. `%26` → `&`). |
| apps/studio/src/server/modules/auth/singpass/singpass.router.ts | Stores `landingUrl` (validated absolute URL) in the Singpass session during `login`, then returns it as `redirectUrl` in the `callback` query. Falls back to `DASHBOARD` when absent. |
| apps/studio/src/features/sign-in/components/EmailLogin/VerificationInput.tsx | After OTP success, passes `callbackUrl` query param along when navigating to the Singpass page, so the URL is not lost between steps. |
| apps/studio/src/components/AuthWrappers/EnforceLoginStatePageWrapper.tsx | Fixes a React anti-pattern: moves the redirect side-effect from render-time (inside `useMemo`) into `useEffect`, and removes the SSR guard since `window` is only accessed inside the effect. |
| apps/studio/src/components/ErrorBoundary/DefaultTrpcError.tsx | Moves side-effects (invalidate, push) into `useEffect` and preserves the current page as `callbackUrl` when redirecting to sign-in. `resetErrorBoundary` is always called after the promise settles (even on a swallowed error), which may hide a failed redirect. |
| apps/studio/src/lib/types/session.ts | Adds optional `landingUrl` field to the Singpass session state; backwards-compatible since `undefined` falls back to `DASHBOARD`. |
| apps/studio/tests/e2e/auth/singpass-callback-url.test.ts | New e2e test verifying the encoded callback URL is correctly forwarded to the server during Singpass login, including the `%26`-preservation case. |
| apps/studio/tests/e2e/smoke.test.ts | Adds a smoke test confirming that accessing a protected page while "logged out" (via localStorage flag) redirects to sign-in with the correct `callbackUrl` query param. |
| apps/studio/src/server/modules/auth/singpass/__tests__/singpass.router.test.ts | Unit tests updated to assert that `landingUrl` is stored in the session on `login` and returned as `redirectUrl` from `callback`. |

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant U as User
    participant B as Browser
    participant OTP as OTP Page
    participant SP as Singpass Page
    participant S as Server (tRPC)
    participant SG as Singpass IdP

    U->>B: Visit /sites/1 (unauthenticated)
    B->>OTP: "Redirect to /sign-in?callbackUrl=%2Fsites%2F1"
    U->>OTP: Enter email + OTP
    OTP->>S: auth.email.verifyOtp
    S-->>OTP: success
    OTP->>SP: "router.push(/sign-in/singpass?callbackUrl=%2Fsites%2F1)"
    Note over SP: getRedirectUrl(router.query)<br/>returns /sites/1 (no double-decode)
    U->>SP: Click "Authenticate with Singpass"
    SP->>S: "auth.singpass.login({ landingUrl: "/sites/1" })"
    Note over S: callbackUrlSchema validates same-origin<br/>Stores landingUrl.toString() in session
    S-->>SP: "{ redirectUrl: authorizationUrl }"
    SP->>SG: redirect to Singpass IdP
    SG-->>B: "callback to /sign-in/singpass/callback?code=..."
    B->>S: "auth.singpass.callback({ code, state })"
    Note over S: Reads landingUrl from session
    S-->>B: "{ redirectUrl: "http://localhost:3000/sites/1" }"
    B->>U: router.replace → /sites/1
```
</details>

<sub>Reviews (9): Last reviewed commit: ["Merge branch &#39;main&#39; into fix/preserve-si..."](https://github.com/opengovsg/isomer/commit/f7c51f44a9d642669881eb9e2a221768722927a4) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44733557)</sub>

### CODE REVIEW

File: apps/studio/src/features/sign-in/components/EmailLogin/VerificationInput.tsx [hunk 1]
```diff
@@ -62,7 +62,12 @@ export const VerificationInput = (): JSX.Element | null => {
   const verifyOtpMutation = trpc.auth.email.verifyOtp.useMutation({
     onSuccess: async () => {
       if (isSingpassEnabled) {
-        await router.push(SIGN_IN_SINGPASS)
+        await router.push({
+          pathname: SIGN_IN_SINGPASS,
+          query: {
+            [CALLBACK_URL_KEY]: router.query[CALLBACK_URL_KEY],
+          },
+        })
```
[the remark this answers is not in the document]
Comment (Benjam11n) [reply apps/studio/src/features/sign-in/components/EmailLogin/VerificationInput.tsx:75]: updated in the latest commit

File: apps/studio/src/server/modules/auth/singpass/__tests__/singpass.router.test.ts [hunk 2]
```diff
@@ -319,12 +323,13 @@ describe("auth.singpass", () => {
...
       // Act
-      await caller.callback({
+      const result = await caller.callback({
         state: JSON.stringify({ state: expect.any(String) }),
         code: "code",
       })
 
       // Assert
+      expect(result.redirectUrl).toBe("http://localhost/sites/1?foo=bar")
       const auditLogs = await db.selectFrom("AuditLog").selectAll().execute()
```
[the remark this answers is not in the document]
Comment (Benjam11n) [reply apps/studio/src/server/modules/auth/singpass/__tests__/singpass.router.test.ts:333]: updated in the latest commit

File: apps/studio/src/server/modules/auth/singpass/__tests__/singpass.router.test.ts [hunk 3]
```diff
@@ -310,6 +313,7 @@ describe("auth.singpass", () => {
           verificationToken: {} as never,
           codeVerifier: "code-verifier",
           nonce: "nonce",
+          landingUrl: "http://localhost/sites/1?foo=bar",
         },
```
[the remark this answers is not in the document]
Comment (Benjam11n) [reply apps/studio/src/server/modules/auth/singpass/__tests__/singpass.router.test.ts:317]: updated in the latest commit

**Review by adriangohjw — CHANGES_REQUESTED** [review]
currently if a unauthenticated user enter a URL e.g. `http://localhost:3000/sites/1/pages/16`, it does not automatically add that to the callbackUrl in the login page, which means they are still redirected to "/" after login in

should auto add that to the callback url to improve UX

though not exactly the issue described in the github issue, i think only make sense to tackle it together in this PR, in order for it to meaningfully improve UX

can i trouble you to do that? thanks!

File: apps/studio/src/server/modules/auth/singpass/singpass.router.ts [hunk 4]
```diff
@@ -92,7 +93,7 @@ export const singpassRouter = router({
         })
       }
 
-      const { codeVerifier, nonce, userId, verificationToken } =
+      const { codeVerifier, nonce, userId, verificationToken, landingUrl } =
         ctx.session.singpass.sessionState
```
[the remark this answers is not in the document]
Comment (Benjam11n) [reply apps/studio/src/server/modules/auth/singpass/singpass.router.ts:98]: this appears to be a pre-existing issue. since this is unrelated to the fix, thoughts on adding this?

**Comment by adriangohjw** [comment]
@Benjam11n the e2e tests are failing, can i trouble you to take a look? thanks!

**Comment by adriangohjw** [comment]
## Code Review (medium effort)

Reviewed the diff against `main`. Two findings, both in `DefaultTrpcError.tsx`:

### 1. `resetErrorBoundary` is no longer guaranteed to run
**File:** `apps/studio/src/components/ErrorBoundary/DefaultTrpcError.tsx:23`

```ts
void router
  .replace(callbackUrlSchema.parse(appendWithRedirect(SIGN_IN, callbackUrl)))
  .then(resetErrorBoundary)
```

Previously `resetErrorBoundary()` was called unconditionally and synchronously. Now it only fires if `router.replace`'s promise resolves. Next.js Pages Router's `push`/`replace` promise is documented to reject with an "Abort fetching component for route" error when a route transition is cancelled (e.g. rapid navigation, or another nav firing before this one settles). Since there's no `.catch`, a rejection means `resetErrorBoundary` never runs — the error boundary stays tripped and the user is stuck on `FullscreenSpinner` with no recovery short of a full reload.

**Suggestion:** add a `.catch` (or `.finally`) that still calls `resetErrorBoundary`.

### 2. Removed cache invalidation on UNAUTHORIZED redirect
**File:** `apps/studio/src/components/ErrorBoundary/DefaultTrpcError.tsx:19`

The `trpc.useUtils()` + `void utils.invalidate()` call that used to run before every UNAUTHORIZED redirect was dropped, with nothing equivalent added.

`apps/studio/src/utils/trpc.ts:190` sets a 10s `staleTime` on the query client. If a user hits UNAUTHORIZED, re-authenticates via the new callback-preserving sign-in flow, and lands back on the same page within that 10s window, React Query may serve previously cached (pre-auth/error-associated) data instead of refetching for the newly authenticated session.

---

The redirect-preservation logic itself (`EnforceLoginStatePageWrapper`, `VerificationInput`, `singpass.router.ts`) checked out end-to-end — encoding, schema validation, and wiring to `SingpassLoginButton`/`SingpassCallback` are all consistent.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by Benjam11n** [comment]
Hi @adriangohjw, just bumping this PR when you have some time. I have addressed the previous feedback. Thanks!
